### PR TITLE
Marketplace 2017.06

### PR DESCRIPTION
**Pull request by Joeper214**
### Changelog

 - Pre release v1.25.0 #360 
-  Pre release v1.26.0 #374 
-  Pre release v1.27.0 #386 
-  Pre release v1.28.0 #391 
-  Pre release v1.29.0 #398 
-  Minor RDS change Scale modal fix #407 

## Features
- Improve Zabbix in Project Prompt Message
- Warning Text on Change scale modal
- Improve DB manangement
- AWSpec test CRUD feature
- CloudFormation Apply Date History
- From TSD to Typings
- Fetch Latest AWS regions
- Add log Notification after yum update execution is successful
- Specify Subnet and VPC on initial deployment
- Reflect ANSI terminal character color/highlighting

## Update process

``` sh
cd YOUR_SKYHOPPER_DIRECTORY
bundle install --path vendor/bundle
bundle exec rake db:migrate RAILS_ENV=production
bundle exec rake db:seed RAILS_ENV=production
cd frontend/
npm i
gulp type
gulp ts
cd ..
```

### CODE REVIEW

File: app/models/database_manager.rb [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+      env = SUFFIX[Rails.env]
+      secrets.each do |key, value|
+        path = Rails.root.join('secrets', "#{key}-#{env}")
+        File.write(path, value)
+      end
+
+      ReloadSecretsJob.perform_now   # runs in Rails process
+      ReloadSecretsJob.perform_later # runs in Sidekiq process
+
+      system("rake db:data:load[#{sqlpath}]")
```
Comment (skyarch-networks) [inline app/models/database_manager.rb]: __[brakeman]__
* Possible command injection [:ref](http://brakemanscanner.org/docs/warning_types/command_injection/)

File: app/models/node.rb [hunk 2]
```diff
@@ -187,36 +187,12 @@ def run_serverspec(infra_id, serverspec_ids, selected_auto_generated)
...
-          'success'
-        else
-          'pending'
-        end
-      else
-        'failed'
-      end
-
+    result =  generate_result(out)
+    Resource.find_by(physical_id: @name).status.servertest.update(value: result[:status_text])
```
Comment (skyarch-networks) [inline app/models/node.rb]: __[brakeman]__
* Model attribute used in file name [:ref](http://brakemanscanner.org/docs/warning_types/file_access/)

File: app/controllers/databases_controller.rb [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+  helper_method :app_setting_is_set
+
+  # POST /databases/export
+  def export
+    authenticate_user!
+
+    time = Time.now.strftime('%Y%m%d%H%M%S')
+    zipfile = DatabaseManager.export_as_zip
+    send_file(zipfile.path, filename: "SkyHopper-db-#{Rails.env}-#{time}.zip")
```
Comment (skyarch-networks) [inline app/controllers/databases_controller.rb]: __[brakeman]__
* Model attribute used in file name [:ref](http://brakemanscanner.org/docs/warning_types/file_access/)

File: app/controllers/servertests_controller.rb [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+
+    @servertest = Servertest.new(infrastructure_id: infra_id)
+    @servertest.value = %!require '<choice_category>spec_helper'\n\n!
+  end
+
+  # GET /servertests/1
+  def show
+    render text: @servertest.value
+  end
+
```
Comment (skyarch-networks) [inline app/controllers/servertests_controller.rb]: __[brakeman]__
* Unescaped model attribute rendered inline [:ref](http://brakemanscanner.org/docs/warning_types/cross_site_scripting/)

File: app/models/servertest_result.rb [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+# This software is released under the MIT License.
+#
+# http://opensource.org/licenses/mit-license.php
+#
+
+class ServertestResult < ActiveRecord::Base
+  belongs_to :resource
+  has_many :servertest_result_details
+  has_many :servertests, through: :servertest_result_details
+  enum status: [:success, :pending, :failed]
```
Comment (skyarch-networks) [inline app/models/servertest_result.rb]: __[rubocop]__
* Style/SymbolArray: Use `%i` or `%I` for an array of symbols. [:ref](https://github.com/bbatsov/ruby-style-guide#percent-i)

File: app/models/database_manager.rb [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+# Copyright (c) 2013-2016 SKYARCH NETWORKS INC.
+#
+# This software is released under the MIT License.
+#
+# http://opensource.org/licenses/mit-license.php
+#
+
+class DatabaseManager
+  SQLPATH = 'tmp/import.sql'.freeze
+  SECRETS = [:secret_key_base, :db_crypt_key].freeze
```
Comment (skyarch-networks) [inline app/models/database_manager.rb]: __[rubocop]__
* Style/SymbolArray: Use `%i` or `%I` for an array of symbols. [:ref](https://github.com/bbatsov/ruby-style-guide#percent-i)

File: app/models/servertest.rb [hunk 7]
```diff
@@ -8,14 +8,15 @@
...
-  has_many :dishes, through: :dish_serverspecs
-  has_many :resource_serverspecs
-  has_many :resources, through: :resource_serverspecs
-  has_many :serverspec_results, through: :serverspec_result_details
+  has_many :dish_servertests
+  has_many :dishes, through: :dish_servertests
+  has_many :resource_servertests
+  has_many :resources, through: :resource_servertests
+  has_many :servertest_results, through: :servertest_result_details
+  enum      category: [:awspec, :serverspec]
```
Comment (skyarch-networks) [inline app/models/servertest.rb]: __[rubocop]__
* Style/SymbolArray: Use `%i` or `%I` for an array of symbols. [:ref](https://github.com/bbatsov/ruby-style-guide#percent-i)

File: app/controllers/app_settings_controller.rb [hunk 8]
```diff
@@ -9,7 +9,7 @@
 class AppSettingsController < ApplicationController
   before_action :authenticate_user!, except: [:show, :create, :chef_create]
 
-  before_action except: [:edit_zabbix, :update_zabbix, :chef_server, :chef_keys, :db, :export_db] do
+  before_action except: [:edit_zabbix, :update_zabbix, :chef_server, :chef_keys] do
```
Comment (skyarch-networks) [inline app/controllers/app_settings_controller.rb]: __[rubocop]__
* Style/SymbolArray: Use `%i` or `%I` for an array of symbols. [:ref](https://github.com/bbatsov/ruby-style-guide#percent-i)

File: app/controllers/servertests_controller.rb [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+# This software is released under the MIT License.
+#
+# http://opensource.org/licenses/mit-license.php
+#
+
+class ServertestsController < ApplicationController
+  include Concerns::InfraLogger
+  class ServertestError < ::StandardError; end
+
+  before_action :set_servertest, only: [:update, :show, :edit, :destroy]
```
Comment (skyarch-networks) [inline app/controllers/servertests_controller.rb]: __[rubocop]__
* Style/SymbolArray: Use `%i` or `%I` for an array of symbols. [:ref](https://github.com/bbatsov/ruby-style-guide#percent-i)

File: app/controllers/servertests_controller.rb [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+  # GET /servertests/results
+  def results
+    physical_id = params.require(:physical_id)
+    infra_id    = params.require(:infra_id)
+    resource = Resource.where(infrastructure_id: infra_id).find_by(physical_id: physical_id)
+
+    @servertest_results = resource.servertest_results.order("created_at desc")
+
+    respond_to do |format|
+      format.json { render json: @servertest_results.as_json(only: [:id, :status, :message, :created_at, :category],
```
Comment (skyarch-networks) [inline app/controllers/servertests_controller.rb]: __[rubocop]__
* Style/SymbolArray: Use `%i` or `%I` for an array of symbols. [:ref](https://github.com/bbatsov/ruby-style-guide#percent-i)

File: app/controllers/servertests_controller.rb [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+        include: [{servertest_result_details: {only: [:id]}},{servertests: {only: [:name, :category]}}, {resource: {only: [:physical_id]}} ]) }
```
Comment (skyarch-networks) [inline app/controllers/servertests_controller.rb]: __[rubocop]__
* Style/SymbolArray: Use `%i` or `%I` for an array of symbols. [:ref](https://github.com/bbatsov/ruby-style-guide#percent-i)

File: spec/jobs/servertest_job_spec.rb [hunk 12]
```diff
@@ -8,38 +8,38 @@
 
 require_relative '../spec_helper'
 
-RSpec.describe ServerspecJob, type: :job do
+RSpec.describe ServertestJob, type: :job do
```
Comment (skyarch-networks) [inline spec/jobs/servertest_job_spec.rb]: __[rubocop]__
* Metrics/BlockLength: Block has too many lines. [76/25]

File: spec/models/servertest_spec.rb [hunk 13]
```diff
@@ -9,25 +9,25 @@
 require_relative '../spec_helper'
 
 # TODO: refactor
-describe Serverspec, type: :model do
-  let(:klass){Serverspec}
+describe Servertest, type: :model do
```
Comment (skyarch-networks) [inline spec/models/servertest_spec.rb]: __[rubocop]__
* Metrics/BlockLength: Block has too many lines. [47/25]

File: spec/controllers/servertests_controller_spec.rb [hunk 14]
```diff
@@ -8,24 +8,25 @@
 
 require_relative '../spec_helper'
 
-describe ServerspecsController, type: :controller do
+describe ServertestsController, type: :controller do
```
Comment (skyarch-networks) [inline spec/controllers/servertests_controller_spec.rb]: __[rubocop]__
* Metrics/BlockLength: Block has too many lines. [340/25]

File: spec/controllers/servertests_controller_spec.rb [hunk 15]
```diff
@@ -8,24 +8,25 @@
...
     before do
-      create_list(:serverspec, 3, infrastructure_id: infrastructure.id)
+      create_list(:servertest, 3, infrastructure_id: infrastructure.id)
       req
     end
 
-    it 'should assign @serverspec' do
-      expect(assigns[:serverspecs]).to be_all{|serverspec|serverspec.kind_of?(Serverspec)}
+    it 'should assign @servertest' do
+      expect(assigns[:servertests]).to be_all{|servertest|servertest.kind_of?(Servertest)}
```
Comment (skyarch-networks) [inline spec/controllers/servertests_controller_spec.rb]: __[rubocop]__
* Lint/AmbiguousBlockAssociation: Parenthesize the param `be_all{|servertest|servertest.kind_of?(Servertest)}` to make sure that the block will be associated with the `be_all` method call. [:ref](https://github.com/bbatsov/ruby-style-guide#syntax)

File: spec/controllers/servertests_controller_spec.rb [hunk 16]
```diff
@@ -45,8 +46,8 @@
         expect(assigns[:infrastructure_name]).to eq infrastructure.stack_name
       end
 
-      it 'should assign @serverspecs where infra id = infra.id(from url)' do
-        expect(assigns[:serverspecs]).to be_all{|serverspec|serverspec.infrastructure_id == infrastructure.id}
+      it 'should assign @servertests where infra id = infra.id(from url)' do
+        expect(assigns[:servertests]).to be_all{|servertest|servertest.infrastructure_id == infrastructure.id}
```
Comment (skyarch-networks) [inline spec/controllers/servertests_controller_spec.rb]: __[rubocop]__
* Lint/AmbiguousBlockAssociation: Parenthesize the param `be_all{|servertest|servertest.infrastructure_id == infrastructure.id}` to make sure that the block will be associated with the `be_all` method call. [:ref](https://github.com/bbatsov/ruby-style-guide#syntax)

File: spec/controllers/servertests_controller_spec.rb [hunk 17]
```diff
@@ -282,29 +283,29 @@
...
 
-    it 'should assign @serverspec_results' do
-      expect(assigns[:serverspec_results]).to eq resource.serverspec_results
+    it 'should assign @servertest_results' do
+      expect(assigns[:servertest_results]).to eq resource.servertest_results
     end
   end
 
-  describe '#run' do
+  describe '#run_serverspec' do
```
Comment (skyarch-networks) [inline spec/controllers/servertests_controller_spec.rb]: __[rubocop]__
* Metrics/BlockLength: Block has too many lines. [63/25]

File: spec/policies/servertest_policy_spec.rb [hunk 18]
```diff
@@ -8,30 +8,30 @@
 
 require_relative '../spec_helper'
 
-describe ServerspecPolicy do
+describe ServertestPolicy do
```
Comment (skyarch-networks) [inline spec/policies/servertest_policy_spec.rb]: __[rubocop]__
* Metrics/BlockLength: Block has too many lines. [51/25]

**Comment by skyarch-networks** [comment]
Click here if you want to create rubocop's auto-correct Pull Request
[Auto Correct by SideCI](https://sideci.com/gh/31754199/issues/pull_request/373/diff/6202e5bc6e24592f334e88e4d7f0bb22d1c85ce2?open_modal=1#rubocop)

See the documentation if you want to get more information about auto-correct.
[Document](https://sideci.com/ja/docs/autofix-rubocop)

File: app/controllers/app_settings_controller.rb [hunk 19]
```diff
@@ -122,11 +142,14 @@ def chef_create
       CfTemplate
       # rubocop:enable Lint/Void
 
+      cf_params = Rails.cache.fetch(CF_PARAMS_KEY) || {}
+      Rails.cache.delete(CF_PARAMS_KEY)
+
```
Comment (skyarch-networks) [inline app/controllers/app_settings_controller.rb]: __[rubocop]__
* Metrics/BlockLength: Block has too many lines. [43/25]

**Comment by skyarch-networks** [comment]
Click here if you want to create rubocop's auto-correct Pull Request
[Auto Correct by SideCI](https://sideci.com/gh/31754199/issues/pull_request/373/diff/7e453178c5ada1ca150396778ce2b34eb668803b?open_modal=1#rubocop)

See the documentation if you want to get more information about auto-correct.
[Document](https://sideci.com/ja/docs/autofix-rubocop)

**Comment by Joeper214** [comment]
Maybe we can release a new version today? @root-beer

**Comment by root-beer** [comment]
It seems this branch has a bug.
`roles: [null, null]`

**Comment by Joeper214** [comment]
Oh,  Can you identify which part is it? Then I'll create a PR for this. Does it occur in other branches ex `develop`?

**Comment by root-beer** [comment]
`develop` is Okay. But `master` has a problem.

**Comment by root-beer** [comment]
I guess it occurs when ridley's version is old.

**Comment by Joeper214** [comment]
maybe because developer has a lot of updates and fixes.

Like this https://github.com/skyarch-networks/skyhopper/pull/397

So, maybe we need to release `pre-release v1.29.0` before releasing a new release in marketplace branch.

**Comment by Joeper214** [comment]
Yeah . the ridley update was the issue.

**Comment by root-beer** [comment]
Sorry. The problem occurs because I'm using Ruby 2.3. This branch works with Ruby 2.2, so there is no problem. :bow:

**Comment by Joeper214** [comment]
But I still encountered the Ridley Error message when trying to view the instance EC2.

So, It is better to set the Ridley update from develop branch before releasing this branch.

**Pull request merged**